internal/querynodev2/delegator/delegator.go

```diff
@@ -684,12 +705,164 @@ func NewShardDelegator(ctx context.Context, collectionID UniqueID, replicaID Uni
+		zap.Int64("topK", topK),
+		zap.Float32("filterRatio", filterRatio))
+
+	var phg commonpb.PlaceholderGroup
+	err := proto.Unmarshal(req.GetReq().GetPlaceholderGroup(), &phg)
+	if err != nil {
+		log.Warn("fail to parse SearchRequest PlaceholderGroup", zap.Error(err))
+		return req, sealeds
+	}
+	if len(phg.GetPlaceholders()) == 0 {
+		return req, sealeds
+	}
+	bSearchVectors := phg.GetPlaceholders()[0].GetValues()
+
+	segments := make([]SegmentEntry, 0)
+	for _, sealed := range sealeds {
+		segments = append(segments, sealed.Segments...)
+	}
+
+	type segmentDistanceStruct struct {
+		segment  SegmentEntry
+		distance float32
+	}
+
+	optimizedSegs := make(map[UniqueID]SegmentEntry, 0)
+	for _, bSearchVector := range bSearchVectors {
+		vectorSegmentDistance := make([]segmentDistanceStruct, 0)
+		for _, segment := range segments {
+			if segment.ClusteringInfos == nil || len(segment.ClusteringInfos.GetVectorClusteringInfos()) == 0 ||
+				segment.ClusteringInfos.GetVectorClusteringInfos()[0].Centroid == nil {
+				optimizedSegs[segment.SegmentID] = segment
+				continue
+			}
+			// currently, segment can contains only one vector cluster
+			vectorClusteringInfo := segment.ClusteringInfos.GetVectorClusteringInfos()[0]
+			distance, err := clustering.CalcVectorDistance(sd.vectorField.dim, sd.vectorField.dataType, bSearchVector, vectorClusteringInfo.Centroid, metricType)
+			if err != nil {
+				log.Warn("Fail to calculate distance between clustering center and search vector, skip this segment", zap.Error(err))
+				optimizedSegs[segment.SegmentID] = segment
+				continue
+			}
+			log.Debug("distance between bSearchVector and cluster center",
+				zap.Int64("segmentID", segment.SegmentID),
+				zap.Float32s("distance", distance),
+				zap.Any("bSearchVector", bSearchVector),
+				zap.Any("clusterCentroid", vectorClusteringInfo.String()))
+			if len(distance) > 0 {
+				vectorSegmentDistance = append(vectorSegmentDistance, segmentDistanceStruct{segment: segment, distance: distance[0]})
+			} else {
+				// if no legal distance is calculated, regard this segment as a normal one.
+				optimizedSegs[segment.SegmentID] = segment
+			}
+		}
+
+		// sort by distance
+		switch metricType {
+		case distance.L2:
+			sort.SliceStable(vectorSegmentDistance, func(i, j int) bool {
+				return vectorSegmentDistance[i].distance < vectorSegmentDistance[j].distance
+			})
+		case distance.IP:
+			// for IP or cosine metric, larger result means more similar, we should sort reverse
+			sort.SliceStable(vectorSegmentDistance, func(i, j int) bool {
+				return vectorSegmentDistance[i].distance > vectorSegmentDistance[j].distance
+			})
+		case distance.COSINE:
+			sort.SliceStable(vectorSegmentDistance, func(i, j int) bool {
+				return vectorSegmentDistance[i].distance > vectorSegmentDistance[j].distance
+			})
+		}
+
+		toFilterSegNum := len(vectorSegmentDistance)
+		targetSegNum := int(float32(toFilterSegNum) * filterRatio)
+		var optimizedRowNums int64
```

Comment:
this seems not a like a strong guarantee.

this should handle similarly like autoindex.

@chasingegg Any suggestion on this?

Reply:
I would suggest still adding re-search logic(#29382) when mvcc is ready #29749 (which is useful for both re-query or re-search). For now, each segment would be touched and topk results num is ensured by index itself, but for clustering, it is possible that skipped segments cause missing results, so re-searching could help on these cases.